### PR TITLE
Reduce verbosity of `clickhouse client --help`

### LINKED ISSUE #42061: ClickHouse CLI --help output is huge
**Describe the issue**
When I run `./clickhouse client --help`, the result doesn't fit on a single screen and overflows the scroll buffer (1000 in my settings).
Such a huge is not the best UX since it's hard to navigate.

**How to reproduce**
* ClickHouse server version: `v22.9` 

**Expected behavior**
I'd expect a smaller output that a human can read.

### PR DESCRIPTION
**Pull request by yariks5s**
Fixes: #42061

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Reduce the verbosity of command line argument `--help` in `clickhouse client` and `clickhouse local`. The previous output is now generated by `--help --verbose`.

### CODE REVIEW

**Comment by robot-clickhouse-ci-1** [comment]
*This is an automated comment for commit 0cc404a74c98077a23de9dae322a8f7687486526 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/62973/0cc404a74c98077a23de9dae322a8f7687486526/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

File: src/Client/ClientBase.cpp [hunk 1]
```diff
@@ -3008,24 +3003,6 @@ void ClientBase::init(int argc, char ** argv)
...
-        const auto & main_options = options_description.main_description->options();
-        std::transform(main_options.begin(), main_options.end(), std::back_inserter(cmd_options), getter);
-    }
-
-    if (options_description.external_description)
-    {
-        const auto & external_options = options_description.external_description->options();
-        std::transform(external_options.begin(), external_options.end(), std::back_inserter(cmd_options), getter);
-    }
-
```
Comment (nikitamikhaylov) [inline src/Client/ClientBase.cpp:3028]: Consider to add another option (something like `--verbose` or `-vvv`) to display all the available settings.

File: src/Client/ClientBase.cpp [hunk 2]
```diff
@@ -2844,13 +2844,33 @@ class OptionsAliasParser
 }
 
 
-void ClientBase::parseAndCheckOptions(OptionsDescription & options_description, po::variables_map & options, Arguments & arguments)
+void ClientBase::verboseHelpMessageIfNeeded(OptionsDescription & options_description)
```
Comment (rschu1ze) [inline src/Client/ClientBase.cpp:2847]: Where exactly is this method called from?

File: programs/benchmark/Benchmark.cpp [hunk 3]
```diff
@@ -598,6 +598,7 @@ int mainEntryClickHouseBenchmark(int argc, char ** argv)
         boost::program_options::options_description desc = createOptionsDescription("Allowed options", getTerminalWidth());
         desc.add_options()
             ("help",                                                            "produce help message")
+            ("verbose",                                                         "print more detailed message")
```
Comment (rschu1ze) [inline programs/benchmark/Benchmark.cpp:601]: `clickhouse-benchmark` is only used for internal testing. We don't really need to "improve" its help output. I guess we don't need to touch Benchmark.cpp at all.

File: src/Client/ClientBase.cpp (hunk 2 above)
Comment (yariks5s) [reply src/Client/ClientBase.cpp:2847]: l. 3059. Just pushed a commit regarding it

File: programs/client/Client.cpp [hunk 4]
```diff
@@ -923,6 +923,7 @@ void Client::printHelpMessage(const OptionsDescription & options_description)
     std::cout << options_description.main_description.value() << "\n";
     std::cout << options_description.external_description.value() << "\n";
     std::cout << options_description.hosts_and_ports_description.value() << "\n";
+    std::cout << "All setting parameters are documented in detail at https://clickhouse.com/docs/ or you can also access them in `system.settings` table.\n\n";
```
Comment (rschu1ze) [inline programs/client/Client.cpp:926]: 1. "in detail" is obsolete
2. the URL is too generic
3. "accessing" settings in system.settings is weird terminology.

--> `All settings are documented at

Comment (rschu1ze) [reply programs/client/Client.cpp:926]: I am getting the following output when I run `clickhouse-client` with `--help`:

```
~/repo/ch1/build/programs (reduce_help_messages =) $ chc --help
Main options:
  --help                              produce help message
  -V [ --version ]                    print version information and exit
  --version-clean                     print version in machine-readable format and exit
  -C [ --config-file ] arg            config-file path
  -q [ --query ] arg                  query; can be specified multiple times (--query "SELECT 1" --query "SELECT 2"...)
  --queries-file arg                  file path with queries to execute; multiple files can be specified (--queries-file file1 file2...)
  -n [ --multiquery ]                 If specified, multiple queries separated by semicolons can be listed after --query. For convenience,
                                      it is also possible to omit --query and pass the queries directly after --multiquery.
  -m [ --multiline ]                  If specified, allow multiline queries (do not send the query on Enter)
  -d [ --database ] arg               database
  --query_kind arg (=initial_query)   One of initial_query/secondary_query/no_query
  --query_id arg                      query_id
  --history_file arg                  path to history file
  --stage arg (=complete)             Request query processing up to specified stage: complete,fetch_columns,with_mergeable_state,with_mer
                                      geable_state_after_aggregation,with_mergeable_state_after_aggregation_and_limit
  --progress [=arg(=tty)] (=default)  Print progress of queries execution - to TTY: tty|on|1|true|yes; to STDERR non-interactive mode:
                                      err; OFF: off|0|false|no; DEFAULT - interactive to TTY, non-interactive is off
  -A [ --disable_suggestion ]         Disable loading suggestion data. Note that suggestion data is loaded asynchronously through a second
                                      connection to ClickHouse server. Also it is reasonable to disable suggestion if you want to paste a
                                      query with TAB characters. Shorthand option -A is for those who get used to mysql client.
  --wait_for_suggestions_to_load      Load suggestion data synchonously.
  -t [ --time ]                       print query execution time to stderr in non-interactive mode (for benchmarks)
  --echo                              in batch mode, print query before execution
  --verbose                           print query and other debugging info
```

There are two visual problems:
1. `--help` and `--verbose` are not next to each other
2. `--verbose` is a weird name, one could easily confuse it with "verbose" output for SELECT statements. What about `--verbose_help`?

File: programs/benchmark/Benchmark.cpp (hunk 3 above)
Comment (yariks5s) [reply programs/benchmark/Benchmark.cpp]: done🙂

File: programs/client/Client.cpp (hunk 4 above)
Comment (yariks5s) [reply programs/client/Client.cpp:926]: > `--verbose` is a weird name, one could easily confuse it with "verbose" output for SELECT statements.

this setting is now used for "verbose" output for SELECT statements and "verbose" output for --help. Wouldn't having two different `verbose` settings for different purposes be confusing?

File: tests/queries/0_stateless/03133_help_message_verbosity.sh [hunk 5]
```diff
@@ -0,0 +1,40 @@
+#!/usr/bin/env bash
+
```
Comment (rschu1ze) [inline tests/queries/0_stateless/03133_help_message_verbosity.sh:2]: I am indifferent if we need a test (I personally would not have written one but I am good with a test).

In any case, what you (generally) want to do is to add a one-line summary of the test purpose. Imagine a future programmer reading this file in a few months from now. The file name could mean many things but not necessarily `clickhouse-client --help`. So the only option to understand what is actually tested is to read the test :-(

Therefore, let's add here something like `# Checks that "clickhouse-client/local --help" prints a brief summary of CLI arguments and "--help --verbose" prints all possible CLI arguments`

File: programs/client/Client.cpp [hunk 6]
```diff
@@ -923,6 +923,7 @@ void Client::printHelpMessage(const OptionsDescription & options_description)
     std::cout << options_description.main_description.value() << "\n";
     std::cout << options_description.external_description.value() << "\n";
     std::cout << options_description.hosts_and_ports_description.value() << "\n";
+    std::cout << "All settings are documented at https://clickhouse.com/docs/en/operations/settings/settings and readable from system.settings.\n\n";
```
Comment (rschu1ze) [inline programs/client/Client.cpp:926]: Sorry, the grammar of my proposal was trash (lol), `[...] and available in system.settings` sounds better.

But to be honest, I would omit the hint to `system.settings`. It kind of duplicates the web documentation and in 99.9% of all cases all settings will take their default anyways (in the other 0.1% of all cases, the user sets them explicitly via CLI or config file but even then one would not need to look into system.settings for that).

Instead, you could perhaps link https://clickhouse.com/docs/en/operations/system-tables/settings from https://clickhouse.com/docs/en/operations/settings/settings (e.g .`All settings on this page are displayed in system.settings`)?

Another problem is that this the sentence reads a bit out-of-place with `--help` (which does not print settings). It actually only makes sense with `--help --verbose`. Is it possible to print it only in the latter case?

File: src/Client/ClientBase.cpp [hunk 7]
```diff
@@ -2956,6 +2956,7 @@ void ClientBase::init(int argc, char ** argv)
     /// Common options for clickhouse-client and clickhouse-local.
     options_description.main_description->add_options()
         ("help", "produce help message")
+        ("verbose", "print query and other debugging info")
```
Comment (rschu1ze) [inline src/Client/ClientBase.cpp:2959]: L. 2958: `("help", "print usage summary, combine with --verbose to display all options")`

File: src/Client/ClientBase.cpp [hunk 8]
```diff
@@ -3008,6 +3008,7 @@ void ClientBase::init(int argc, char ** argv)
 
     addOptions(options_description);
 
+    OptionsDescription options_description_not_verbose = options_description;
```
Comment (rschu1ze) [inline src/Client/ClientBase.cpp:3011]: Cosmetics: `options_description_non_verbose`

\+ newline after it

File: src/Client/ClientBase.cpp [hunk 9]
```diff
@@ -3042,11 +3043,17 @@ void ClientBase::init(int argc, char ** argv)
...
 
+    if (options.count("verbose"))
+        config().setBool("verbose", true);
+
     /// Output of help message.
     if (options.count("help")
         || (options.count("host") && options["host"].as<std::string>() == "elp")) /// If user writes -help instead of --help.
     {
-        printHelpMessage(options_description);
+        if (config().getBool("verbose", false))
```
Comment (rschu1ze) [inline src/Client/ClientBase.cpp:3053]: I don't think we need to provide a default value (`false`).

File: programs/client/Client.cpp (hunk 4 above)
Comment (rschu1ze) [reply programs/client/Client.cpp]: Yeah. Let's retain a single `--verbose` then (as currently implemented)

File: src/Client/ClientBase.cpp (hunk 9 above)
Comment (yariks5s) [reply src/Client/ClientBase.cpp:3053]: The config parameter current is available only when this parameter is used. That means we have two solutions:

1. set the value of this config to false if we cannot find the `verbose` in the list of used settings (not really common for our code)
2. leave the default value to false if we cannot find `verbose` in the list of used settings

Comment (rschu1ze) [reply src/Client/ClientBase.cpp:3053]: Okay, let's keep things then as is.

File: programs/client/Client.cpp (hunk 6 above)
Comment (yariks5s) [reply programs/client/Client.cpp:926]: So, if I understood correctly, we need to remove the hint for `system.settings`. Add a link to `system.settings` docs for https://clickhouse.com/docs/en/operations/settings/settings and try to show the hint when the user uses only the verbose version?

File: tests/queries/0_stateless/03133_help_message_verbosity.sh (hunk 5 above)
Comment (yariks5s) [reply tests/queries/0_stateless/03133_help_message_verbosity.sh:2]: That's a really good point, will add the comment, thanks

File: programs/client/Client.cpp (hunk 6 above)
Comment (rschu1ze) [reply programs/client/Client.cpp:926]: yes

**Review by rschu1ze — APPROVED** [review]

**Pull request merged**